### ISSUE TITLE
Query view - Find & Replace bug

### ISSUE THREAD

**Issue by Kiersm** [sequelpro/sequelpro#2442]
Values in 'Replace' dropdown do not take effect.

Steps to reproduce:
1) In Query View, Cmd+F to open find/replace dialog
2) Perform a find and replace using any two values
3) Perform a second with different values (so there's history to choose from)
4) Perform a 3rd find/replace. This time, select the replace value for step 2 from the list. When 'Replace' is clicked, the value from step 3 is inserted instead.

I think it's an issue with change handling on the replace dropdown. If you manually edit the value the value in the field is used correctly. But simply selecting, doesn't update the value being used, even though the correct value shows in the dropdown.

**Comment by Allendar** [comment]
The call to catch this is in `SPCustomQuery.m` in the call `willChangeSelectionFromCharacterRange`. This isn't actually a bug. What you experienced was a mere accident of behavior: the Replace button will replace what it has selected, or where (in range) the cursor is. So when you say Replace but haven't done a Next/Previous or still have text selected from a previous replace, it will replace that which is selected. The `willChangeSelectionFromCharacterRange` method enum clearly states a position and a range. When no length is given, it will just pop in the text and move all the characters to the right. Otherwise it will replace all characters in length with the characters you are giving up in your "Replace with:" box.

Issue can be closed.

**Comment by Kiersm** [comment]
Sorry - no. It's not a case of the wrong selection being replaced, but rather the correct select is being replaced with the wrong text. Despite changing the value in the replace dropdown, the previous value was used, not the newly chosen one. If you type the new replace value it's fine, but simply switching the value in the dropdown field does not.

**Comment by Allendar** [comment]
You have a point. I'll try to test what happens there later today. The Find & Replace itself is native OS X bound tho, so it has to have to do with something that is either buffered or updated wrongly when delegates on that UITextView are triggered.

**Comment by Allendar** [comment]
Hmm it depends strongly on who you interpret Replace/Next/Find/Find&Replace. Can you re-iterate your steps even more specific? Like literally which buttons you pressed on each action. I tried many variations, yet all behavior seems normal to me.

Replace & Find is a weird thing, because when you assume you're still in your "replacement session" but you change your replace value, it won't mind that it will replace a matched value of your earlier search. So simply put the Find & Replace tool is a bit more simplistic than you would (intuitively) expect.

**Comment by Kiersm** [comment]
Hi, the steps in the original post should cover it. It's essentially that rather than typing the replace value, you select it from the dropdown (which means you need a history of replace values built up to select from). I've just retried it and it reproduces the error.

**Comment by Kiersm** [comment]
As soon as you type into the replace value field (even just add a character onto the end of the selected value) it works fine. That makes me think that simply selecting a value from the replace dropdown doesn't trigger an update to whatever variable controls what the replace value is.

**Comment by Allendar** [comment]
Yes the replace feature is direct on where you are in that place in time. Hitting Replace will not jump to the first result first and then Replace. It will instantly replace what you have selected (or where simply your cursor is at that moment).

**Comment by Kiersm** [comment]
That's not what I'm saying... I do have the string to replace selected (highlighted) in the query editor.

The issue is that, in the find / replace dialog, if you select a replace value from drop down as opposed to pasting / typing it into the replace field, the replace button does NOT replace the selected text with the correct value

**Comment by Kiersm** [comment]
![untitled-2](https://cloud.githubusercontent.com/assets/8314873/15098034/79aa87f2-1528-11e6-8300-b05cd5769548.jpg)
In this screenshot, I performed a F&R to change [sometext] to 'ReplaceMe'. I then highlighted that, did a new F&R and selected 'abc' in the replace with dropdown.

Clicking Replace does nothing (i.e., it's still inserting 'ReplaceMe' which was the previous value of the Replace With field)

**Comment by Kiersm** [comment]
A further example:

![untitled-3](https://cloud.githubusercontent.com/assets/8314873/15098043/dea1689c-1528-11e6-84e3-9517dde0b3d7.jpg)

Here, I asked it to replace the text at the end of the paragraph. After clicking 'replace', the result is shown below:

![untitled-4](https://cloud.githubusercontent.com/assets/8314873/15098044/debf3638-1528-11e6-9911-afa202f06858.jpg)

As you can see, it is still using 'ReplaceMe' as the replacement value because I selected 'abc' from the dropdown, rather than typing it. Without a keypress event, the value the software uses to replace the selected text does not get updated.
